### ISSUE TITLE
Incorrect out axis order with xmap when using grad(dot)

### ISSUE THREAD

**Issue by sharadmv** [google/jax#6333]
It looks like a weird combination of using collectives with grad(dot) causes an out axis to be in the incorrect position. This could be an issue with vmap as well.

In this repro `user` and `movie` are two mapping axes: 
```python
import jax
from jax import random, lax
from jax.experimental import maps
d = 2

def run(key, w):
  u_key, m_key, w_key = random.split(key, 3) # unmapped
  u_key = random.fold_in(u_key, lax.axis_index('user')) # [user]
  m_key = random.fold_in(m_key, lax.axis_index('movie')) # [movie]
  w_key = random.fold_in(
      random.fold_in(w_key, lax.axis_index('user')), lax.axis_index('movie')) # [user, movie]
  u, m  = random.normal(u_key, [d]), random.normal(m_key, [d]) # ([user, d], [movie, d])
  def func(m):
    return u.dot(m) * w # [user, movie]
  out = u * m # [user, movie, d]
  gs = jax.grad(func)(m) # [user, movie, d]

  out = lax.psum(out, 'user') # [movie, d]
  gs = lax.psum(gs, 'user') # [movie, d]
  return out, gs

W = random.normal(random.PRNGKey(0), [32, 16])
out = maps.xmap(run,
  in_axes=[{}, {0: 'user', 1: 'movie'}],
  out_axes=(['movie', ...], ['movie', ...])
  )(random.PRNGKey(0), W)
assert out[0].shape == (16, 2), out[0].shape # PASSES
assert out[1].shape == (16, 2), out[1].shape  # FAILS (2, 16)
```

If I define `func` as follows (no dot):
```python
def func(m):
    return (u * m).sum()
```
the assertion passes. On the other hand, with this `func`:
```python
def func(m):
    return (u.dot(m) * u * m).sum()
```
the assertion fails.

It appears to work with vanilla `vmap`:
```python
run2 = jax.vmap(jax.vmap(run, in_axes=(None, 0), axis_name='movie', out_axes=0),
                in_axes=(None, 0), axis_name='user', out_axes=None)
out = run2(random.PRNGKey(0), W)
assert out[0].shape == (16, 2), out[0].shape # PASSES
assert out[1].shape == (16, 2), out[1].shape  # PASSES
```

Also, it works in Python3.6 but fails in Python3.7; @mattjj suggested that it might have to do with dictionary ordering.

EDIT: it actually appears to be nondeterministic in both python3.6 and python3.7, so I don't really understand what is going on :(.

**Comment by sharadmv** [comment]
Actually, this doesn't repro at head and is likely an artifact of a patch I made for another shape issue. I'll close this

**Comment by sharadmv** [comment]
Actually I think it does repro at head. Here's a colab to confirm: https://colab.research.google.com/drive/1jh6Cfo9m_rLWjcbRj9qeS28i7L_57owB?resourcekey=0-PDyFLDU3Y7LVQsEpq8UfqA&usp=sharing

**Comment by sharadmv** [comment]
@apaszke Any clue what is happening here? Thanks for the help!

**Comment by sharadmv** [comment]
This no longer repros at HEAD possibly fixed by #6564 (thanks @apaszke!). I'll close this issue.
